### ISSUE TITLE
Slow exec error when no /etc/group

### ISSUE THREAD

**Issue by rumpl** [moby/moby#46903]
### Description

Trying to exec into a container that doesn't have `/etc/group` shows an error but then just lingers there for 10 seconds:

### Reproduce

```console
root@b4f40d17b559:/go/src/github.com/docker/docker# ID=$(docker run -d docker/desktop-kubernetes-pause:3.7)
Unable to find image 'docker/desktop-kubernetes-pause:3.7' locally
3.7: Pulling from docker/desktop-kubernetes-pause
aff472d3f83e: Download complete
Digest: sha256:bb6ed397957e9ca7c65ada0db5c5d1c707c9c8afc80a94acbe69f3ae76988f0c
Status: Downloaded newer image for docker/desktop-kubernetes-pause:3.7
root@b4f40d17b559:/go/src/github.com/docker/docker# docker ps
CONTAINER ID   IMAGE                                 COMMAND    CREATED         STATUS         PORTS     NAMES
d005f8037c37   docker/desktop-kubernetes-pause:3.7   "/pause"   5 seconds ago   Up 5 seconds             musing_meninsky
root@b4f40d17b559:/go/src/github.com/docker/docker# time docker exec -it -u 0:root $ID /bin/sh
unable to find group root: no matching entries in group file


real    0m10.046s
user    0m0.025s
sys     0m0.018s
```

### Expected behavior

Expected more something like this:

```console
root@b4f40d17b559:/go/src/github.com/docker/docker# ID=$(docker run -d docker/desktop-kubernetes-pause:3.7)^C
root@b4f40d17b559:/go/src/github.com/docker/docker# docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
root@b4f40d17b559:/go/src/github.com/docker/docker# ID=$(docker run -d docker/desktop-kubernetes-pause:3.7)
root@b4f40d17b559:/go/src/github.com/docker/docker# time docker exec -it $ID /bin/sh
unable to find group root: no matching entries in group file

real    0m0.067s
user    0m0.017s
sys     0m0.011s
```

### docker version

```bash
master
```

### docker info

```bash
master
```

**Comment by vvoland** [comment]
The 10 second timeout comes from:
https://github.com/moby/moby/blob/debcb769393dd31cb1569b0cf5a80e0985fa5c6d/daemon/resize.go#L40-L61

It doesn't happen if you use `-d`:
```bash
$ ID=$(docker run -d docker/desktop-kubernetes-pause:3.7)
$ time docker exec -dit $ID /bin/sh
Error response from daemon: open /var/lib/docker/rootfs/overlayfs/e4ed54e27b236d511975d609c8c435bcc0168782e999ca842e0bee41e0256b6a/etc/group: no such file or directory

real    0m0.019s
user    0m0.010s
sys     0m0.006s
```


For some reason, the reason is not returned on exec start and the CLI already is already at a later stage where it monitors the tty resize signals:
https://github.com/docker/cli/blob/cddd1869a8d1f9e0f1350e805340c09f65daf81c/cli/command/container/exec.go#L202-L206

**Comment by vvoland** [comment]
Hmm, looks like the API returns the error differently with `Detach`:

https://github.com/moby/moby/blob/debcb769393dd31cb1569b0cf5a80e0985fa5c6d/api/server/router/container/exec.go#L152-L159

**Comment by thaJeztah** [comment]
Is it executing the group lookup (`GIDFromPath`)? https://github.com/containerd/containerd/blob/194a1fdd2cde35bc019ef138f30485e27fe0913e/oci/spec_opts.go#L1069-L1083

Also wondering if that should special handling for "not found" errors for those files.


There's also another possibility; is there a codepath where something may end up trying `getent` ? https://github.com/moby/moby/blob/debcb769393dd31cb1569b0cf5a80e0985fa5c6d/pkg/idtools/idtools_unix.go#L103-L116

**Comment by vvoland** [comment]
Yes, `WithAppendAdditionalGroups` explicitly reads the `/etc/group` and errors out if it's not accessible:
https://github.com/containerd/containerd/blob/194a1fdd2cde35bc019ef138f30485e27fe0913e/oci/spec_opts.go#L884-L897

Fun fact, it uses a function from `moby/sys` to parse the groups file:
https://github.com/moby/sys/blob/4950d7687cf6c9b138dc0e18c2c7351e1f6ed497/user/user.go#L156-L164

**Comment by thaJeztah** [comment]
> Fun fact, it uses a function from moby/sys to parse the groups file:

Yes, that's the code that was originally in runc, and was moved to a separate module

**Comment by thaJeztah** [comment]
I'm also curious what it's trying to look up in this case; It looks like `GIDFromPath` unconditionally reads `/etc/group`, then takes the first entry it finds as `group` and uses its ID;

```bash
docker run --rm alpine cat /etc/group
root:x:0:root
bin:x:1:root,bin,daemon
...
```

So even if no user is set, it's reading those files to (surprise!) find out that it must run as `0` (`root`)?

Which makes me curious;

- MUST the first entry always be for `root` ?
- Is it possible / allowed to have a different group ID for the first entry? (i.e., `root` does not have `0` as group?)

All of that at least assumes the container is running a "distro" image, and does not take `from scratch` images into account;

- if no user is defined (`""`); should it just bail out and use `0`?
- if `/etc/group` and/or `/etc/password` are not present, should it just use `0` for both? (i.e.: `from scratch` containers)
- for numeric "user" or "group", should it even look-up anything? (there was some debate bout that, and possibly that was changed already as part of a "CVE" (fun fact: Linux userland tools DO allow a user-name or group-name that's a number))

**Comment by rumpl** [comment]
Opened a PR in containerd https://github.com/containerd/containerd/pull/9494

**Comment by thaJeztah** [comment]
- opened https://github.com/moby/moby/issues/46917 for my comment above

**Comment by rumpl** [comment]
Updated this issue: removed the `containerd-integration` label because it happens with graph drivers too

**Comment by thaJeztah** [comment]
We should also look at that error-message; I don't think the absolute path of the `/etc/group` file is very informative to the user;

```bash
printf 'FROM alpine\nRUN rm -f /etc/group' | docker build -t nogroup -
ID=$(docker run -dit nogroup)

time docker exec -it -u 0:root $ID echo hello
open /var/lib/docker/rootfs/overlayfs/d5b1024abe8f3ce9208d5b9612ef56e1e5cdd0bf706a434a5caf4ddbf016da73/etc/group: no such file or directory

docker exec -it -u 0:root $ID echo hello  0.04s user 0.05s system 0% cpu 10.165 total
```
